WV-861 add next page UI [TEAM REVIEW]

itcreativeusa opened the pull request:
1)Implemented backend logic to support pagination with total_count, items_per_page, start_index, and end_index.
2) Added previous_page_url and next_page_url for navigation.
3) Conditionally hid pagination UI when items fit on a single page.
4)Updated the frontend to display "Previous" and "Next" links with the current page number.

itcreativeusa committed:
[WV-861]Add 'Next Page' UI on Politician Admin page

itcreativeusa committed:
[WV-861]Add 'Next Page' UI on Politician Admin page

[WV-861]Add 'Next Page' UI on Politician Admin page

DaleMcGrew commented on templates/politician/politician_list.html:
```diff
@@ -336,5 +367,6 @@ <h1>Politicians</h1>
     });
 
 </script>
+<script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
```
Hi Katerenya, we shouldn't have to include this here because jquery is already included in WeVoteServer/templates/template_base.html on line 24. You can upgrade that version in that file if you need a newer version.

DaleMcGrew commented on politician/views_admin.py:
```diff
@@ -1166,7 +1169,34 @@ def politician_list_view(request):
+    
     template_values = {
+        "start_index": start_index,
+        'politician_list': politician_list,
```
Please add these template variables in alphabetical order, because that helps us make sure we don't duplicate existing variables. (I see "politician_list" and "politician_search" for example are already passed in further down in this template_values dict.)

DaleMcGrew commented on politician/views_admin.py:
```diff
@@ -1166,7 +1169,34 @@ def politician_list_view(request):
+    end_index = start_index + items_per_page
+
+    # Fetch only the items for the current page
+    politician_list = politician_query.order_by("politician_name")[start_index:end_index]
```
We have already retrieved politician_list above, and augmented it with data from the Representative table. This replaces that work with a brand new query. I think this pagination logic should be included up around lines 992-996 where the current retrieve is happening:

        if not positive_value_exists(show_all):
            politician_list = politician_query.order_by('politician_name')[:25]
        else:
            # We still want to limit to 200
            politician_list = politician_query.order_by('politician_name')[:200]

DaleMcGrew commented on templates/politician/politician_list.html:
```diff
@@ -51,22 +51,24 @@ <h1>Politicians</h1>
 
 <form name="state_code_form" method="get" action="{% url 'politician:politician_list' %}">
     {% csrf_token %}
+    <input type="hidden" name="politician_search" value="{{ politician_search|default_if_none:'' }}" />
```
politician_search is already being included in the form down below on your line 74 in an input, and doesn't need to be added here as a hidden value.

DaleMcGrew commented on templates/politician/politician_list.html:
```diff
@@ -51,22 +51,24 @@ <h1>Politicians</h1>
 <form name="state_code_form" method="get" action="{% url 'politician:politician_list' %}">
     {% csrf_token %}
+    <input type="hidden" name="politician_search" value="{{ politician_search|default_if_none:'' }}" />
+    <input type="hidden" name="show_all" value="{{ show_all|default_if_none:0 }}" />
```
show_all is already included below as a checkbox, and doesn't need to be included here as a hidden value.

DaleMcGrew commented on templates/politician/politician_list.html:
```diff
@@ -51,22 +51,24 @@ <h1>Politicians</h1>
     {% csrf_token %}
+    <input type="hidden" name="politician_search" value="{{ politician_search|default_if_none:'' }}" />
+    <input type="hidden" name="show_all" value="{{ show_all|default_if_none:0 }}" />
+    <input type="hidden" name="exclude_politician_analysis_done" value="{{ exclude_politician_analysis_done|default_if_none:0 }}" />
```
exclude_politician_analysis_done is already included below as a checkbox, and doesn't need to be included here as a hidden value.

DaleMcGrew reviewed (changes requested):
Hi Katerenya, overall you are on the right track. I have these comments for your next iteration. Thank you!

itcreativeusa committed:
[WV-861] Update politician_list.html and views_admin.py

politician/views_admin.py and. templates/politician/politician_list.html

politician/politician_list.html:
Removed redundant hidden inputs and unnecessary jQuery script inclusion.

- Removed duplicate hidden input fields for `politician_search`, `show_all`, and `exclude_politician_analysis_done` since these are already included elsewhere in the form as input fields or checkboxes.
- Removed the redundant jQuery script inclusion, as it is already included in `WeVoteServer/templates/template_base.html`.
Refactored pagination logic and updated template variables.
Views_admin.py:
- Moved pagination logic to align with the initial `politician_list` retrieval around lines 992-996.
- Ensured `politician_list` is properly paginated without replacing the augmented data from the Representative table.
- Added pagination-related template variables (`has_previous_page`, `has_next_page`, `previous_page_url`, `next_page_url`) in alphabetical order to the `template_values` dictionary to prevent duplication.

itcreativeusa commented:
https://github.com/user-attachments/assets/682c1faa-1ef2-4587-9173-1432f1c8c577

itcreativeusa committed:
fix reloading to page 1 if state changed or checkbox

fix reloading to page 1 if state changed or checkbox

itcreativeusa commented:
This last screen recording shows my last commit : [fix reloading to page 1 if state changed or checkbox]. 
 With these changes, It doesn't go to page 1 or reload if the state was changed or the checkbox checked.

https://github.com/user-attachments/assets/0f62cb7e-03ba-489e-8cab-f73a52b616a2

DaleMcGrew commented:
👍 Thank you for working through all of the challenges of this page!
